subsys/compress/src/lzma.c

```diff
@@ -0,0 +1,263 @@
+
+/* Assume the maximum LZMA dictionary size to limit the RAM buffer size for the decompressed
+ * stream.
+ */
+#define MAX_LZMA_DICT_SIZE  (128 * 1024)
+
+/* Header size for lzma2 */
+#define LZMA2_HEADER_SIZE 2
+
+#if !defined(CONFIG_NRF_COMPRESS_LZMA_VERSION_LZMA1) && \
+	!defined(CONFIG_NRF_COMPRESS_LZMA_VERSION_LZMA2)
+#error "Missing selection of lzma algorithm selection, please select " \
+	"CONFIG_NRF_COMPRESS_LZMA_VERSION_LZMA1 or CONFIG_NRF_COMPRESS_LZMA_VERSION_LZMA2"
+#endif
+
+static void *lzma_probs_alloc(ISzAllocPtr p, size_t size)
+{
+#if defined(CONFIG_NRF_COMPRESS_MEMORY_TYPE_STATIC)
+	static uint16_t lzma_probs[MAX_LZMA_PROB_SIZE];
+
+	if (size > sizeof(lzma_probs)) {
+		LOG_ERR("Compress library tried to allocate too large a buffer (0x%x)", size);
+		return NULL;
+	}
+
+	return lzma_probs;
+#else
+	void *buffer = malloc(size);
+
+	if (buffer == NULL) {
+		LOG_ERR("Failed to allocate nRF compression library buffer (0x%x)", size);
+	}
+
+	return buffer;
+#endif
+}
+
+static void lzma_probs_free(ISzAllocPtr p, void *address)
+{
+#if defined(CONFIG_NRF_COMPRESS_MEMORY_TYPE_MALLOC)
+	free(address);
+#endif
+}
+
+const static ISzAlloc lzma_probs_allocator = {
+	.Alloc = lzma_probs_alloc,
+	.Free = lzma_probs_free,
+};
+
+#if defined(CONFIG_NRF_COMPRESS_MEMORY_TYPE_STATIC)
+static uint8_t lzma_dict[MAX_LZMA_DICT_SIZE];
+#else
+static uint8_t *lzma_dict = NULL;
+#endif
+
+static bool allocated_probs = false;
+#ifdef CONFIG_NRF_COMPRESS_LZMA_VERSION_LZMA2
+static CLzma2Dec lzmaDec;
+#else
+static CLzmaDec lzmaDec;
+#endif
+
+static int lzma_reset(void);
+
+static int lzma_init(void)
+{
+	int rc = 0;
+
+#if defined(CONFIG_NRF_COMPRESS_MEMORY_TYPE_MALLOC)
+	if (lzma_dict != NULL) {
+		/* Already allocated */
+		lzma_reset();
+
+		return rc;
+	}
+
+	lzma_dict = (uint8_t *)malloc(MAX_LZMA_DICT_SIZE);
+
+	if (lzma_dict == NULL) {
+		rc = -ENOMEM;
+	}
+#endif
+
+	return rc;
+}
+
+static int lzma_deinit(void)
+{
+#if defined(CONFIG_NRF_COMPRESS_MEMORY_TYPE_MALLOC)
+	if (lzma_dict != NULL) {
+		free(lzma_dict);
+		lzma_dict = NULL;
+	}
+#endif
+
+	if (allocated_probs) {
+		allocated_probs = false;
+#ifdef CONFIG_NRF_COMPRESS_LZMA_VERSION_LZMA2
+		Lzma2Dec_FreeProbs(&lzmaDec, &lzma_probs_allocator);
+#else
+		LzmaDec_FreeProbs(&lzmaDec, &lzma_probs_allocator);
+#endif
+	}
+
+	return 0;
+}
+
+static int lzma_reset(void)
+{
+	if (allocated_probs) {
+		allocated_probs = false;
+#ifdef CONFIG_NRF_COMPRESS_LZMA_VERSION_LZMA2
+		Lzma2Dec_FreeProbs(&lzmaDec, &lzma_probs_allocator);
+#else
+		LzmaDec_FreeProbs(&lzmaDec, &lzma_probs_allocator);
+#endif
+	}
+
+	return 0;
+}
+
+static int lzma_bytes_needed(void)
+{
+#ifdef CONFIG_NRF_COMPRESS_LZMA_VERSION_LZMA2
+	return (allocated_probs ? CONFIG_NRF_COMPRESS_CHUNK_SIZE : LZMA2_HEADER_SIZE);
+#else
+	return (allocated_probs ? CONFIG_NRF_COMPRESS_CHUNK_SIZE : LZMA_PROPS_SIZE);
+#endif
+}
+
+static int lzma_decompress(const uint8_t *input, size_t input_size, bool last_part,
```

Comment:
I guess this can not decompress two streams in parallel as they will share the same dictionary?

Reply:
correct, single instance

Reply:
One stream is enough for our needs.
Limitation should be documented.

Reply:
I have added a placeholder parameter for potential multiple use in future which for now is just NULL. Will add limitation of single stream to the rst files when subsystem is documented